Tools/Scripts/libraries/resultsdbpy/resultsdbpy/model/test_context.py

```diff
@@ -81,6 +84,52 @@ def recurse(partial_test, results):
         for key, value in results.items():
             recurse(key, value)
 
+    @classmethod
+    def get_test_result_status(cls, test_result_json, will_filter_expected=False):
+        failure_type = None
+        failure_number = None
+        if test_result_json.get('stats', None):
+            if test_result_json.get('start_time', None):
+                failure_number = test_result_json['stats']['tests{}failed'.format('_unexpected_' if will_filter_expected else '_')]
+            else:
+                failure_number = test_result_json['stats']['worst_tests{}failed'.format('_unexpected_' if will_filter_expected else '_')]
+            if 'worst_tests_run' in test_result_json and test_result_json['stats']['worst_tests_run'] <= 1:
+                failure_number = None
+
+            for type in Expectations.STATS_FAILURE_TYPES:
+                if test_result_json['stats'].get('tests{}{}'.format('_unexpected_' if will_filter_expected else '_', type.lower()), 0) > 0:
+                    failure_type = type
+        else:
+            resultId = Expectations.string_to_state_ids(test_result_json['actual'])
+            if will_filter_expected:
+                resultId = Expectations.string_to_state_ids(cls.unexpected_results(test_result_json['actual'], test_result_json['expected']))
+            for type in Expectations.FAILURE_TYPES:
+                if Expectations.string_to_state_ids(type) >= resultId:
+                    failure_type = type
+        return failure_type, failure_number
+
+    @classmethod
+    def unexpected_results(cls, results, expectations):
+        r = results.split('.')
+        for expectation in expectations.split(' '):
+            try:
+                i = r.index(expectation)
+                r = r[:i] + r[i + 1:]
+            except ValueError:
+                pass
+            if expectation == 'FAIL':
+                for expectation in ['TEXT', 'AUDIO', 'IMAGE']:
```

Comment:
This line feels like the wrong way to do this.

I think what we really want is to define a sort of bucket where `FAIL` can mean any of `FAIL`, `TEXT`, `AUDIO` or `IMAGE`. The same generally applies to `PASS` and `WARNING`. I don't have an _exact_ conception of this code, but it feels like it should use the next-worst bucket as a boundary condition. ie, everything between `FAIL` and `TIMEOUT` is in the `FAIL` bucket.

The reason I recommend this is because this file's conception of failure modes deliberately allows for us to insert new failure modes we haven't thought about yet in-between existing ones.

Reply:
This piece of code is a python translation for https://github.com/WebKit/WebKit/blob/main/Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/expectations.js#L57

Reply:
Should probably use variable names when possible, then, and provide a comment pointing to that code (and comment in that code pointing here). If we were doing this in Python exclusively, we'd try and be more clever. Understandable that you would want to basically copy existing JS code, though.